### PR TITLE
[WIP] Interface all the things...

### PR DESCRIPTION
**Pull request by david-driscoll**
**Description:**
Let me prefix this with I really wish we didn't have to do this, but looking at the referenced issues I don't see the root problem being solved by the TypeScript compiler anytime soon.

There are problems when using `rxjs` with tools like `npm link` that cause the compiler to fail to resolve types properly and create lots of compiler errors, as it sees two copies of the same class, and they are not compatible.

The goal of this change is to do the following:
* Make the public `Observable` an interface.
* Make other public classes interfaces as well (`Subscription`, `Subscriber`, etc).
* Make sure `Observable` is still the same `Observable` to external consumers.

**Related issue (if exists):**
https://github.com/Microsoft/TypeScript/issues/6496
https://github.com/Microsoft/TypeScript/issues/7755
https://github.com/ReactiveX/rxjs/issues/1744

**Issues**
* Exporting new types will require jumping through similar interface hoops that this PR will solve.

**Work Inprogress Todos**
* [ ] Rename `Obsevable` -> `ObservableImpl`
* [ ] Rename `IObservable` -> `Observable`
* [ ] Create `ObservableStatic` interface
* [ ] Fixup TSLint errors
* [ ] Update tests to use new interfaces (so they compile!)
* [ ] Repeat renames for other interfaces such as `Subscription` (anything that the consumer may want to expose publicly)

cc @blesh @robwormald @kwonoj @staltz 

cc @mhegazy @ahejlsberg - maybe this can be fixed easily in the compiler?  :pray:   

As it stands `rxjs` (and any other library for that matter) cannot export any `class`es without causing potential compiler issues down the line for others.  Essentially we're being penalized because we're using TypeScript to develop the library.

Are there other simpler fixes?  Can the compiler emit an identifier in the `.d.ts` files that so that it knows "hey this is the same interface as the other one I have over here"?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -65,7 +65,7 @@
     "build_test": "shx rm -rf ./dist/ && npm-run-all build_cjs clean_spec build_spec test_mocha",
     "build_cover": "shx rm -rf ./dist/ && npm-run-all build_cjs build_spec cover",
     "build_docs": "npm-run-all build_es6_for_docs build_global build_spec tests2png decision_tree_widget && esdoc -c esdoc.json",
-    "build_spec": "tsc --project ./spec --pretty",
+    "build_spec": "tsc --project ./spec",
```
Comment (saneyuki) [inline package.json]: @david-driscoll why did you remove `--pretty` flag?

Comment (david-driscoll) [inline package.json]: It was to get errors to be detected by vscode, this will be restored once the branch gets ready for land.

Comment (saneyuki) [inline package.json]: aha. thank you

**Comment by david-driscoll** [comment]
Closing this for now, as there may be an alternative... opening an issue for that.
